dandiapi/api/services/permissions/dandiset.py

```diff
@@ -0,0 +1,118 @@
+    existing_owner_set = set(existing_owners)
+    new_owner_set = set(users)
+
+    # Delete all existing owners
+    DandisetUserObjectPermission.objects.filter(
+        content_object=dandiset.pk, permission__codename='owner'
+    ).delete()
+
+    # Set owners to new list
+    for user in users:
+        add_dandiset_owner(dandiset, user)
+
+    # Return the owners added/removed so they can be emailed
+    removed_owners = existing_owner_set - new_owner_set
+    added_owners = new_owner_set - existing_owner_set
+    return removed_owners, added_owners
+
+
+def is_dandiset_owner(dandiset: Dandiset, user: AbstractBaseUser | AnonymousUser) -> bool:
+    if isinstance(user, AnonymousUser):
+        return False
+
+    user = typing.cast(User, user)
+    return user.has_perm('owner', dandiset)
+
+
+def is_owned_asset(asset: Asset, user: AbstractBaseUser | AnonymousUser) -> bool:
+    """Return `True` if this asset belongs to a dandiset that the user is an owner of."""
+    if user.is_anonymous:
+        return False
+
+    user = typing.cast(User, user)
+    asset_dandisets = Dandiset.objects.filter(versions__in=asset.versions.all())
+    asset_dandisets_owned_by_user = DandisetUserObjectPermission.objects.filter(
+        content_object__in=asset_dandisets,
+        user=user,
+        permission__codename='owner',
+    )
+
+    return asset_dandisets_owned_by_user.exists()
+
+
+def get_owned_dandisets(
+    user: AbstractBaseUser | AnonymousUser,
+    include_superusers=True,  # noqa: FBT002
+) -> QuerySet[Dandiset]:
+    return get_objects_for_user(user, 'owner', Dandiset, with_superuser=include_superusers)
+
+
+def get_visible_dandisets(user: AbstractBaseUser | AnonymousUser) -> QuerySet[Dandiset]:
+    """
+    Return a queryset containing all dandisets visible to the given user.
+
+    This is basically all dandisets except embargoed dandisets not owned by the given user.
+    """
+    # We would like to do something like:
+    # Dandiset.filter(embargo_status=OPEN | permission__owned)
+    # but this is not possible with django-guardian; the `get_objects_for_user` shortcut must
+    # be used instead.
+    # We would like to do something like:
+    # queryset = Dandiset.objects.filter(embargo_status=OPEN).union(get_objects_for_user(...))
+    # but you cannot filter or perform many other common queryset operations after using
+    # union().
+    # Therefore, we resort to fetching the list of all primary keys for dandisets owned by the
+    # current user with `get_objects_for_user(...)`, then filter for either those dandisets or
+    # OPEN dandisets. There aren't very many dandisets and most users won't own very many of
+    # them, so there shouldn't be too many dandisets in the list.
+
+    owned_dandiset_pks = get_owned_dandisets(user).values('pk').all()
+    return Dandiset.objects.filter(
+        Q(embargo_status=Dandiset.EmbargoStatus.OPEN) | Q(pk__in=owned_dandiset_pks)
+    ).order_by('created')
```

Comment:
This comment is incorrect, this actually can be done with one query -
```suggestion
def get_visible_dandisets(user: AbstractBaseUser | AnonymousUser) -> QuerySet[Dandiset]:
    """
    Return a queryset containing all dandisets visible to the given user.

    This is basically all dandisets except embargoed dandisets not owned by the given user.
    """

    # Get dandisets owned by this user, potentially including embargoed dandisets
    owned_dandisets = Dandiset.objects.filter(
        pk__in=(
            DandisetUserObjectPermission.objects.filter(
                user=user, permission__codename='owner'
            ).values('pk'),
        )
    )

    open_dandisets = Dandiset.objects.filter(embargo_status=Dandiset.EmbargoStatus.OPEN)

    # The union of these two sets is the set of dandisets visible to the user
    return owned_dandisets.union(open_dandisets)
```

Although I see now that this was copied from another file and introduced in this PR, so we can also defer it to a follow up if you prefer.

Reply:
It seems like implementing this causes some tests and endpoints to fail, due to the union operation preventing much else from happening to a queryset. It actually seems like that was captured in the comment:

```
    # We would like to do something like:
    # queryset = Dandiset.objects.filter(embargo_status=OPEN).union(get_objects_for_user(...))
    # but you cannot filter or perform many other common queryset operations after using
    # union().
```

Although I imagine you're referring to the first part of that comment that mentions django-guardian. If we can clean this up, I'd prefer to do it in a follow-on PR. Although I'd like to continue to stay away from `union` if we can.

Reply:
Sounds good. I think we could avoid `union`, but we can defer further discussion to a future PR

Reply:
@jjnesbitt or @mvandenburgh, could one of you file an issue to capture this?